### ISSUE TITLE
Default FastqParser from bioparser is very slow

### ISSUE THREAD

**Issue by SamStudio8** [clara-genomics/racon-gpu#14]
Hi folks,
I've been running racon-gpu with good results, so much so that I've now noticed that wall time performance is hampered not by the polishing, but by the parsing of reads. I've tested this theory by writing a small standalone program that involves a bioparser::FastqParser on one of our `fq.gz` files, and confirmed that it can take up to 30 minutes to read one of our Zymo mock community FASTQ files (~15GB).

I tried to diagnose the underlying parser itself without that much luck. I'm pretty sure the performance bottleneck is something to do with the handling of various small(ish) buffers that are used to hold the file buffer and the sequence/qual buffers. In the end, I thought it would instead be quicker to just implement Heng Li's excellent [readfq](https://github.com/lh3/readfq/) library as another template in `bioparser`.

As a proof of concept, I've managed to implement `readfq` in my fork of `bioparser` as `HLFastqParser`. I've patched my version of `racon-gpu` to use this parser instead, with a reasonable performance improvement:

Before:
```
[racon::Polisher::] total = 3207.377 s
```

After:
```
[racon::Polisher::] total = 870.203 s
```

The caveat to this is I haven't done much advanced C++ programming (such as templating), so my implementation ignores the parameters given to it by the `Parser` interface (such as how much memory to use), and just does the absolute minimum required to be a valid parser. I am happy to polish (pun intended) this implementation if you would like it to be merged to the bioparser upstream.

**Comment by vellamike** [comment]
Hi Sam, 

Glad you're finding it useful - I've also noticed that the parser is slow, we would be very interested in a fix for this because it's also going to be an issue on our other projects.

Out of interest, have you tried parsing a `fq` file and comparing it to the corresponding `fq.gz`?

Very interesting to see the massive speedup you got from getting @lh3 's parser into Bioparser - @rvaser do you have any thoughts on the best way to get these changes into bioparser?

**Comment by SamStudio8** [comment]
@vellamike Hey Mike, yes it was also much faster to use the uncompressed FQ. I thought at first that this was a zlib performance issue, but I think it's more to do with the various buffers in `FastqParser`. My C++ foo wasn't good enough to give @rvaser's code a proper lookover, so I just ported @lh3's parser instead. It's still quicker to use the `HLFastqParser` when the file is uncompressed too, but the improvement is less dramatic (ie. ~25seconds vs. ~60+seconds).

**Comment by rvaser** [comment]
Hi Sam, Mike,
not sure what is the cause of such long parsing times. I thought that zlib had problems with `fseek`, because I am reading in chunks of 1GB at a time in Racon. Can you please try replacing `kChunkSize` at https://github.com/isovic/racon/blob/master/src/polisher.cpp#L226 and https://github.com/isovic/racon/blob/master/src/polisher.cpp#L308 with `-1` and see the total time for both zipped and not zipped file?

**Comment by lh3** [comment]
I have just updated readfq/kseq.h to the version used in minimap2. That version fixes a rare bug.

On FASTQ parsing in general, zlib decompression is indeed slower than kseq.h, but a not-so-careful parser can be slower than decompression.

> I thought that zlib had problems with fseek

gzseek() jumps to the start of the file and decompresses everything up to the requested offset. It also breaks streaming, so should be avoided.

**Comment by rvaser** [comment]
Thanks for the clarification Heng. I will reorganize bioparser fully over the weekend.

**Comment by SamStudio8** [comment]
Thanks @lh3, I've updated my `bioparser` repo accordingly. 
@rvaser Let me know if you want a PR of my attempt to implement `HLFastqParser` as a starting point?

**Comment by rvaser** [comment]
I'll let you know @SamStudio8, thanks.

**Comment by rvaser** [comment]
@SamStudio8 & @vellamike, I have updated bioparser accordingly (racon as well, although at new adress: https://github.com/lbcb-sci/racon). Sorry for the delay :/

**Comment by SamStudio8** [comment]
Hey @rvaser, thanks for that! I've just pulled your latest version and given it a try. It is indeed much faster now and operates on the same order of magnitude as `HLFastqParser`. Any idea what the issue was? I suspected the memory buffers, but I couldn't be sure.

For what it's worth @lh3's implementation still slightly outperforms (I'm seeing ~250s vs. ~150s for the sequence parsing step) on my 15G mock community, so I guess it's up to you whether you'd be interested in a PR for `HLFastqParser` in the future.

**Comment by rvaser** [comment]
Hi @SamStudio8, the issue was that I was using `gzseek` when parsing in chunks (i.e. 1GB at a time) which takes time as pointed out by @lh3. In addition, I replaced copying char by char into the bigger buffer with `memcpy` which is often vectorized.

I run some tests on a 30GB FASTQ file (both compressed and uncompressed). When compressed, bioparser needs ~280s, kseq needs ~260s, and when the file is uncompressed, the times are ~60s and ~40s, respectively. At this point I would like to keep it as is, will see in the future.

Thanks all for pointing out the performance issue! :)

**Comment by SamStudio8** [comment]
@vellamike Given that @rvaser's patch brings the parse time down to about that of kseq, we could bump the version of bioparser in racon-gpu/CGA and close this.

**Comment by vellamike** [comment]
> @SamStudio8 & @vellamike, I have updated bioparser accordingly (racon as well, although at new adress: https://github.com/lcbc-sci/racon). Sorry for the delay :/

This address is giving me a 404. Is the repo private?

**Comment by rvaser** [comment]
I am a moron, it is https://github.com/lbcb-sci/racon.

**Comment by vellamike** [comment]
@rvaser is https://github.com/lbcb-sci/racon the "official" repo going forward? We should update our fork if that's the case.

Thanks for these improvments to bioparser - this is fantastic. Do we know know why kseq still outperforms bioparser?

Also, it seems like @SamStudio8 is seeing a bigger difference between bioparser and kseq, do we know why?

**Comment by rvaser** [comment]
@vellamike, yes that is the new official repository. Not sure where the speed gain is in kseq.
